contribs/gmf/src/controllers/abstract.js

```diff
@@ -651,4 +655,149 @@ gmf.AbstractController.prototype.search_ = function(query, overlay) {
     });
 };
 
+function isIOS() {
+  return /iPad|iPhone|iPod/.test(navigator.userAgent) &&
+    !window['MSStream'];
+}
+
+// Get heading depending on devices
+function headingFromDevices(deviceOrientation) {
+  let hdg = deviceOrientation.getHeading();
+  let orientation = window.orientation;
+  if (hdg === undefined) {
+    return undefined;
+  }
+  if (angular.isDefined(hdg) && !isIOS()) {
+    hdg = -hdg;
+    if (window.screen.orientation.angle) {
+      orientation = window.screen.orientation.angle;
+    }
+  }
+  switch (orientation) {
+    case -90:
+      hdg = hdg - (Math.PI / 2);
+      break;
+
+    case 180:
+      hdg = hdg + Math.PI;
+      break;
+
+    case 90:
+      hdg = hdg + (Math.PI / 2);
+      break;
+
+    case 270:
+      hdg = hdg - (Math.PI / 2);
+      break;
+
+    default:
+      hdg = hdg;
+  }
```

Comment:
one switch with 5 entry is clearly worst!
and you can split it to be more readable like:
```js
// Normalize to be between -90 and 180 
orientation =  ((orientation + 179) % 360 - 179);
// Add to hdg in radian
hdg += orientation * Math.PI / 180
```